vm/systemSmartContracts/liquidStaking.go

```diff
@@ -0,0 +1,244 @@
+	defer l.mutExecution.RUnlock()
+
+	err := CheckIfNil(args)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if !l.flagLiquidStaking.IsSet() {
+		l.eei.AddReturnMessage("liquid staking contract is not enabled")
+		return vmcommon.UserError
+	}
+
+	switch args.Function {
+	case core.SCDeployInitFunctionName:
+		return l.init(args)
+	case "claimDelegatedPosition":
+		return l.claimDelegatedPosition(args)
+	case "claimRewardsFromPosition":
+		return l.claimRewardsFromDelegatedPosition(args)
+	case "reDelegateRewardsFromPosition":
+		return l.reDelegateRewardsFromPosition(args)
+	case "unDelegateWithPosition":
+		return l.unDelegateWithPosition(args)
+	case "returnPosition":
+		return l.returnPosition(args)
+	}
+
+	l.eei.AddReturnMessage(args.Function + " is an unknown function")
+	return vmcommon.UserError
+}
+
+func (l *liquidStaking) init(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, l.liquidStakingSCAddress) {
+		l.eei.AddReturnMessage("invalid caller")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		l.eei.AddReturnMessage("not a payable function")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) != 1 {
+		l.eei.AddReturnMessage("invalid number of arguments")
+		return vmcommon.UserError
+	}
+	tokenID := args.Arguments[0]
+	l.eei.SetStorage([]byte(tokenIDKey), tokenID)
+
+	return vmcommon.Ok
+}
+
+func (l *liquidStaking) getTokenID() []byte {
+	return l.eei.GetStorage([]byte(tokenIDKey))
+}
+
+func (l *liquidStaking) checkArgumentsWhenPositionIsInput(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.ESDTTransfers) < 1 {
+		l.eei.AddReturnMessage("function requires liquid staking input")
+		return vmcommon.UserError
+	}
+	if args.CallValue.Cmp(zero) != 0 {
+		l.eei.AddReturnMessage("function is not payable in eGLD")
+		return vmcommon.UserError
+	}
+	for _, esdtTransfer := range args.ESDTTransfers {
+		if !bytes.Equal(esdtTransfer.ESDTTokenName, l.getTokenID()) {
+			l.eei.AddReturnMessage("wrong liquid staking position as input")
+			return vmcommon.UserError
+		}
+	}
+	err := l.eei.UseGas(uint64(len(args.ESDTTransfers)) * l.gasCost.MetaChainSystemSCsCost.LiquidStakingOps)
+	if err != nil {
+		l.eei.AddReturnMessage(err.Error())
+		return vmcommon.OutOfGas
+	}
+
+	return vmcommon.Ok
+}
+
+func (l *liquidStaking) claimDelegatedPosition(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if args.CallValue.Cmp(zero) != 0 {
+		l.eei.AddReturnMessage("function is not payable in eGLD")
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) == 0 {
+		l.eei.AddReturnMessage("not enough arguments")
+		return vmcommon.UserError
+	}
+	if len(args.ESDTTransfers) > 0 {
+		l.eei.AddReturnMessage("function is not payable in ESDT")
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (l *liquidStaking) claimRewardsFromDelegatedPosition(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
```

Comment:
claimRewardsViaLiquidStaking is already implemented in delegation.go. This method shouldn't be also implemented?

Reply:
methods from delegation.go can be called by liquidStaking contract only

Reply:
ok